src/AzureAuth/Commands/Ado/CommandPat.cs

```diff
@@ -63,16 +98,104 @@ private enum OutputMode
         {
-            logger.LogInformation("coming soon");
+            var accessToken = this.AccessToken(publicClientAuth, eventData);
+            if (accessToken == null)
+            {
+                logger.LogError("Failed to acquire an Azure DevOps access token. Re-run with '--verbosity debug' for more info.");
+                return 1;
+            }
+
+            var cache = this.Cache();
+            var client = this.Client(accessToken.Token);
+            var manager = new PatManager(cache, client);
+
+            using (new CrossPlatLock(LockfilePath))
+            {
+                var pat = manager.GetPatAsync(this.PatOptions()).Result;
+
+                // Do not use logger to avoid printing PATs into log files.
+                Console.WriteLine(FormatPat(pat, this.Output));
+            }
+
             return 0;
         }
+
+        private static string FormatPat(PatToken pat, OutputMode output) => output switch
+        {
+            OutputMode.None => string.Empty,
+            OutputMode.Status => $"\"{pat.DisplayName}\" valid until {pat.ValidTo:O}",
+            OutputMode.Token => pat.Token,
+            OutputMode.Base64 => pat.Token.Base64(),
+            OutputMode.Header => pat.Token.AsHeader(AzureAuth.Ado.Authorization.Basic),
+            OutputMode.HeaderValue => pat.Token.AsHeaderValue(AzureAuth.Ado.Authorization.Basic),
+            OutputMode.Json => pat.AsJson(),
+            _ => throw new ArgumentOutOfRangeException(nameof(output)),
+        };
+
+        private TokenResult AccessToken(IPublicClientAuth publicClientAuth, CommandExecuteEventData eventData)
+        {
+            return publicClientAuth.Token(
+                AzureAuth.Ado.Constants.AdoParams,
+                AccessTokenAuthModes,
+                this.Domain,
+                AccessTokenPrompt,
+                AccessTokenTimeout,
+                eventData);
+        }
+
+        private PatOptions PatOptions()
+        {
+            return new PatOptions
+            {
+                Organization = this.Organization,
+                DisplayName = this.DisplayName,
+                Scopes = this.Scopes,
+            };
+        }
+
+        private IPatClient Client(string accessToken)
+        {
+            var baseUrl = new Uri(string.Join('/', AzureAuth.Ado.Constants.BaseUrl, this.Organization));
+            var credentials = new VssOAuthAccessTokenCredential(accessToken);
+            var connection = new VssConnection(baseUrl, credentials);
+            var tokensHttpClientWrapper = new TokensHttpClientWrapper(connection);
+            return new PatClient(tokensHttpClientWrapper);
+        }
+
+        private IPatCache Cache()
+        {
+            var storageProperties = new StorageCreationPropertiesBuilder(
+                PatStorageParameters.CacheFileName,
+                CacheDirectory)

```

Comment:
It's certainly a related value, but it's not _quite_ constant, IMO. It's both operating system and user-dependent. For example, given a user `foobar` they might have the following cache directories
- Windows: `C:\Users\foobar\AppData\Local\Programs\AzureAuth`
- MacOS: `/Users/foobar/.azureauth`
- Linux: `/home/foobar/.azureauth`

After being defined once at runtime they never change again for the lifetime of the application, but they're not as rigidly defined as other "true constants" in my mind.

@kyle-rader-msft discussed adding this value to its own static class and memoizing the value. What do you think about that, @mvanchaa?

Reply:
I agree with Tucker here. Since we use some logic to create them and then they are constants from there on, they are not truly constants. But great point, @mvanchaa :) It's almost who you talk to and a preference thing.

Reply:
I posit that `static readonly` from a use perspective is a "Constant" that can be initialized. So I'm fine with moving it to (a new) `Microsoft.Authentication.AzureAuth.Constants` file and calling it `AppDirectory` or something...

Reply:
Done.